File: sdks/csharp/src/KeyAuth/KeyAuthClient.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace KeyAuth.Sdk;

/// <summary>
/// KeyAuth SaaS C# 客户端 SDK
///
/// 面向终端软件的客户端 SDK，封装 9 个验证 API：
/// Login / Verify / Heartbeat / Bind / Unbind / GetVar / Notice / Version / Logout
///
/// 依赖：.NET 6+ + System.Text.Json（无第三方包）
///
/// 签名算法（与后端 internal/middleware/signature.go 一致）：
///     原文 = METHOD\nPATH?QUERY\nTIMESTAMP\nNONCE\nBODY
///     签名 = HMAC-SHA512/256(secret, 原文) → 64 位小写 hex
///
/// 注：.NET 原生不直接提供 HMACSHA512/256；优先尝试 BouncyCastle 提供者；
///     若不可用则回退 HMACSHA256（与 Python/PHP SDK 回退策略一致）。
///     生产环境建议安装 BouncyCastle.Cryptography NuGet 包以启用 SHA-512/256。
///
/// 铁律 04：API 地址 / AppKey / SignSecret 由调用方传入，SDK 内不硬编码
/// 铁律 06：所有接口错误抛出 KeyAuthException(code, message)，不静默吞异常
/// </summary>
public class KeyAuthClient
{
    private const string Version = "0.4.0";

    private readonly string _apiBase;
    private readonly string _appKey;
    private readonly string _signSecret;
    private readonly HttpClient _httpClient;
    private static readonly SHA512_256Provider _shaProvider = new();

    public KeyAuthClient(string apiBase, string appKey, string signSecret)
    {
        _apiBase = apiBase.EndsWith("/") ? apiBase[..^1] : apiBase;
        _appKey = appKey;
        _signSecret = signSecret;
        _httpClient = new HttpClient { Timeout = TimeSpan.FromSeconds(10) };
    }

    // ==================== 公共 API ====================

    /// <summary>Login 登录（首次自动绑定设备）</summary>
    public Task<JsonElement> LoginAsync(string cardKey, string hwid, string deviceName = "", string deviceType = "")
        => PostAsync("/api/v1/client/login", new
        {
            card_key = cardKey,
            hwid = hwid,
            device_name = deviceName,
            device_type = deviceType
        });

    /// <summary>Verify 验证卡密有效性</summary>
    public Task<JsonElement> VerifyAsync(string cardKey, string hwid)
        => PostAsync("/api/v1/client/verify", new { card_key = cardKey, hwid = hwid });

    /// <summary>Heartbeat 心跳保活</summary>
    public Task<JsonElement> HeartbeatAsync(string cardKey, string hwid)
        => PostAsync("/api/v1/client/heartbeat", new { card_key = cardKey, hwid = hwid });

    /// <summary>Bind 手动绑定设备</summary>
    public Task<JsonElement> BindAsync(string cardKey, string hwid, string deviceName = "", string deviceType = "")
        => PostAsync("/api/v1/client/bind", new
        {
            card_key = cardKey,
            hwid = hwid,
            device_name = deviceName,
            device_type = deviceType
        });

    /// <summary>Unbind 解绑设备</summary>
    public Task<JsonElement> UnbindAsync(string cardKey, string hwid)
        => PostAsync("/api/v1/client/unbind", new { card_key = cardKey, hwid = hwid });

    /// <summary>GetVar 获取云变量</summary>
    public Task<JsonElement> GetVarAsync(string cardKey, string varKey)
        => PostAsync("/api/v1/client/get_var", new { card_key = cardKey, var_key = varKey });

    /// <summary>Notice 获取应用公告</summary>
    public Task<JsonElement> NoticeAsync()
        => PostAsync("/api/v1/client/notice", new { });

    /// <summary>Version 检查版本更新</summary>
    public Task<JsonElement> VersionAsync(string currentVersion = "", string platform = "")
        => PostAsync("/api/v1/client/version", new { current_version = currentVersion, platform = platform });

    /// <summary>Logout 退出登录</summary>
    public Task<JsonElement> LogoutAsync(string cardKey, string hwid)
        => PostAsync("/api/v1/client/logout", new { card_key = cardKey, hwid = hwid });

    // ==================== 内部方法 ====================

    private async Task<JsonElement> PostAsync(string path, object payload)
    {
        string body = JsonSerializer.Serialize(payload);
        string timestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds().ToString();
        string nonce = RandomNonce();

        // 签名原文：METHOD\nPATH?QUERY\nTIMESTAMP\nNONCE\nBODY
        string signString = $"POST\n{path}\n{timestamp}\n{nonce}\n{body}";
        string signature = _shaProvider.HmacHex(_signSecret, signString);

        using var req = new HttpRequestMessage(HttpMethod.Post, _apiBase + path)
        {
            Content = new StringContent(body, Encoding.UTF8, "application/json")
        };
        req.Headers.Add("X-App-Key", _appKey);
        req.Headers.Add("X-Timestamp", timestamp);
        req.Headers.Add("X-Nonce", nonce);
        req.Headers.Add("X-Signature", signature);

        HttpResponseMessage resp;
        try
        {
            resp = await _httpClient.SendAsync(req);
        }
        catch (Exception e)
        {
            throw new KeyAuthException(1006, "网络请求失败: " + e.Message, 0);
        }

        string raw = await resp.Content.ReadAsStringAsync();
        using var doc = JsonDocument.Parse(raw);
        var root = doc.RootElement;

        int code = root.TryGetProperty("code", out var c) && c.ValueKind == JsonValueKind.Number
            ? c.GetInt32()
            : -1;
        string message = root.TryGetProperty("message", out var m) && m.ValueKind == JsonValueKind.String
            ? m.GetString() ?? "未知错误"
            : "未知错误";

        if ((int)resp.StatusCode != 200 || code != 0)
        {
            throw new KeyAuthException(code, message, (int)resp.StatusCode);
        }

        if (root.TryGetProperty("data", out var data))
        {
            return data.Clone();
        }
        return default;
    }

    private static string RandomNonce()
    {
        Span<byte> b = stackalloc byte[16];
        RandomNumberGenerator.Fill(b);
        return Convert.ToHexString(b).ToLowerInvariant();
    }

    public static string GetVersion() => Version;
}

/// <summary>
/// SHA-512/256 HMAC 提供者
/// 优先使用 BouncyCastle（若运行时已加载）；否则回退 .NET 原生 HMACSHA256
/// </summary>
internal sealed class SHA512_256Provider
{
    private readonly bool _useBouncyCastle;

    public SHA512_256Provider()
    {
        // 反射探测 BouncyCastle.Cryptography 是否已加载
        _useBouncyCastle = Type.GetType("Org.BouncyCastle.Crypto.Digests.Sha512_256Digest, BouncyCastle.Cryptography") != null;
    }

    public string HmacHex(string secret, string msg)
    {
        byte[] key = Encoding.UTF8.GetBytes(secret);
        byte[] data = Encoding.UTF8.GetBytes(msg);
        byte[] raw = _useBouncyCastle ? HmacBouncyCastle(key, data) : HmacFallback(key, data);
        return Convert.ToHexString(raw).ToLowerInvariant();
    }

    private static byte[] HmacFallback(byte[] key, byte[] data)
    {
        using var h = new HMACSHA256(key);
        return h.ComputeHash(data);
    }

    private static byte[] HmacBouncyCastle(byte[] key, byte[] data)
    {
        // 通过反射调用 BouncyCastle HMac + Sha512_256Digest，避免硬依赖 NuGet 包
        var asm = Type.GetType("Org.BouncyCastle.Crypto.Digests.Sha512_256Digest, BouncyCastle.Cryptography")!;
        var digest = Activator.CreateInstance(asm)!;
        var macType = Type.GetType("Org.BouncyCastle.Crypto.Macs.HMac, BouncyCastle.Cryptography")!;
        var mac = Activator.CreateInstance(macType, digest)!;
        var keyParamType = Type.GetType("Org.BouncyCastle.Crypto.Parameters.KeyParameter, BouncyCastle.Cryptography")!;
        var keyParam = Activator.CreateInstance(keyParamType, key)!;
        macType.GetMethod("Init")!.Invoke(mac, new[] { keyParam });
        macType.GetMethod("BlockUpdate", new[] { typeof(byte[]), typeof(int), typeof(int) })!
            .Invoke(mac, new object[] { data, 0, data.Length });
        var result = new byte[32];
        macType.GetMethod("DoFinal", new[] { typeof(byte[]), typeof(int) })!
            .Invoke(mac, new object[] { result, 0 });
        return result;
    }
}

/// <summary>KeyAuth API 错误</summary>
public class KeyAuthException : Exception
{
    public int Code { get; }
    public new string Message { get; }
    public int HttpStatus { get; }

    public KeyAuthException(int code, string message, int httpStatus) : base($"[{code}] {message}")
    {
        Code = code;
        Message = message;
        HttpStatus = httpStatus;
    }
}

File: sdks/java/src/main/java/com/keyauth/sdk/KeyAuthClient.java
package com.keyauth.sdk;

import com.fasterxml.jackson.databind.ObjectMapper;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * KeyAuth SaaS Java 客户端 SDK
 *
 * <p>面向终端软件的客户端 SDK，封装 9 个验证 API：
 * login / verify / heartbeat / bind / unbind / getVar / notice / version / logout</p>
 *
 * <p>依赖：JDK 11+（java.net.http.HttpClient）+ Jackson Databind（JSON 处理）</p>
 *
 * <p>签名算法（与后端 internal/middleware/signature.go 一致）：
 * <pre>
 * 原文 = METHOD\nPATH?QUERY\nTIMESTAMP\nNONCE\nBODY
 * 签名 = HMAC-SHA512/256(secret, 原文) → 64 位小写 hex
 * </pre>
 * 注：JDK 17+ 支持 HmacSHA512/256，JDK 11-16 回退 HmacSHA256（与 Python/PHP SDK 回退策略一致）</p>
 *
 * <p>铁律 04：API 地址 / AppKey / SignSecret 由调用方传入，SDK 内不硬编码<br>
 * 铁律 06：所有接口错误抛出 KeyAuthException(code, message)，不静默吞异常</p>
 */
public class KeyAuthClient {

    private static final String VERSION = "0.4.0";

    private final String apiBase;
    private final String appKey;
    private final String signSecret;
    private final HttpClient httpClient;
    private final ObjectMapper mapper;
    private final SecureRandom random;

    /**
     * 构造客户端
     *
     * @param apiBase    后端 API 根地址（如 https://yourdomain.com）
     * @param appKey     应用 AppKey（ak_ 开头）
     * @param signSecret 应用 SignSecret（sk_ 开头，AES 解密后的明文）
     */
    public KeyAuthClient(String apiBase, String appKey, String signSecret) {
        this.apiBase = apiBase.endsWith("/") ? apiBase.substring(0, apiBase.length() - 1) : apiBase;
        this.appKey = appKey;
        this.signSecret = signSecret;
        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(5))
                .build();
        this.mapper = new ObjectMapper();
        this.random = new SecureRandom();
    }

    // ==================== 公共 API ====================

    /** Login 登录（首次自动绑定设备） */
    public Map<String, Object> login(String cardKey, String hwid, String deviceName, String deviceType) throws KeyAuthException {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("card_key", cardKey);
        payload.put("hwid", hwid);
        payload.put("device_name", deviceName);
        payload.put("device_type", deviceType);
        return post("/api/v1/client/login", payload);
    }

    /** Verify 验证卡密有效性（不绑定，不增加使用次数） */
    public Map<String, Object> verify(String cardKey, String hwid) throws KeyAuthException {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("card_key", cardKey);
        payload.put("hwid", hwid);
        return post("/api/v1/client/verify", payload);
    }

    /** Heartbeat 心跳保活 */
    public Map<String, Object> heartbeat(String cardKey, String hwid) throws KeyAuthException {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("card_key", cardKey);
        payload.put("hwid", hwid);
        return post("/api/v1/client/heartbeat", payload);
    }

    /** Bind 手动绑定设备 */
    public Map<String, Object> bind(String cardKey, String hwid, String deviceName, String deviceType) throws KeyAuthException {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("card_key", cardKey);
        payload.put("hwid", hwid);
        payload.put("device_name", deviceName);
        payload.put("device_type", deviceType);
        return post("/api/v1/client/bind", payload);
    }

    /** Unbind 解绑设备（扣时 UnbindDeductSeconds） */
    public Map<String, Object> unbind(String cardKey, String hwid) throws KeyAuthException {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("card_key", cardKey);
        payload.put("hwid", hwid);
        return post("/api/v1/client/unbind", payload);
    }

    /** GetVar 获取云变量 */
    public Map<String, Object> getVar(String cardKey, String varKey) throws KeyAuthException {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("card_key", cardKey);
        payload.put("var_key", varKey);
        return post("/api/v1/client/get_var", payload);
    }

    /** Notice 获取应用公告 */
    public Map<String, Object> notice() throws KeyAuthException {
        return post("/api/v1/client/notice", new LinkedHashMap<>());
    }

    /** Version 检查版本更新 */
    public Map<String, Object> version(String currentVersion, String platform) throws KeyAuthException {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("current_version", currentVersion);
        payload.put("platform", platform);
        return post("/api/v1/client/version", payload);
    }

    /** Logout 退出登录 */
    public Map<String, Object> logout(String cardKey, String hwid) throws KeyAuthException {
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("card_key", cardKey);
        payload.put("hwid", hwid);
        return post("/api/v1/client/logout", payload);
    }

    // ==================== 内部方法 ====================

    private Map<String, Object> post(String path, Map<String, Object> payload) throws KeyAuthException {
        String body;
        try {
            body = mapper.writeValueAsString(payload);
        } catch (Exception e) {
            throw new KeyAuthException(1006, "JSON 编码失败: " + e.getMessage(), 0);
        }

        String timestamp = String.valueOf(System.currentTimeMillis() / 1000);
        String nonce = randomNonce();

        // 签名原文：METHOD\nPATH?QUERY\nTIMESTAMP\nNONCE\nBODY
        String signString = String.join("\n", "POST", path, timestamp, nonce, body);
        String signature = hmacSHA512_256Hex(signSecret, signString);

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(apiBase + path))
                .header("Content-Type", "application/json")
                .header("X-App-Key", appKey)
                .header("X-Timestamp", timestamp)
                .header("X-Nonce", nonce)
                .header("X-Signature", signature)
                .timeout(Duration.ofSeconds(10))
                .POST(HttpRequest.BodyPublishers.ofString(body, StandardCharsets.UTF_8))
                .build();

        HttpResponse<String> resp;
        try {
            resp = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        } catch (Exception e) {
            throw new KeyAuthException(1006, "网络请求失败: " + e.getMessage(), 0);
        }

        Map<String, Object> envelope;
        try {
            envelope = mapper.readValue(resp.body(), Map.class);
        } catch (Exception e) {
            String preview = resp.body();
            if (preview != null && preview.length() > 200) {
                preview = preview.substring(0, 200);
            }
            throw new KeyAuthException(1006, "响应非 JSON: " + preview, resp.statusCode());
        }

        Number code = (Number) envelope.get("code");
        int codeVal = code != null ? code.intValue() : -1;
        if (resp.statusCode() != 200 || codeVal != 0) {
            String msg = (String) envelope.getOrDefault("message", "未知错误");
            throw new KeyAuthException(codeVal, msg, resp.statusCode());
        }

        Object data = envelope.get("data");
        if (data instanceof Map) {
            @SuppressWarnings("unchecked")
            Map<String, Object> dataMap = (Map<String, Object>) data;
            return dataMap;
        }
        return new LinkedHashMap<>();
    }

    /** HMAC-SHA512/256 → 64 位小写 hex（与后端 crypto.HMACSHA256 对齐） */
    static String hmacSHA512_256Hex(String secret, String msg) throws KeyAuthException {
        try {
            // 优先尝试 HmacSHA512/256（JDK 17+ 支持）
            Mac mac;
            try {
                mac = Mac.getInstance("HmacSHA512/256");
            } catch (Exception e) {
                // 回退 HmacSHA256（JDK 11-16，与 Python/PHP SDK 回退策略一致）
                mac = Mac.getInstance("HmacSHA256");
            }
            mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), mac.getAlgorithm()));
            byte[] raw = mac.doFinal(msg.getBytes(StandardCharsets.UTF_8));
            return toHex(raw);
        } catch (Exception e) {
            throw new KeyAuthException(1006, "签名计算失败: " + e.getMessage(), 0);
        }
    }

    private String randomNonce() {
        byte[] b = new byte[16];
        random.nextBytes(b);
        return toHex(b);
    }

    private static String toHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder(bytes.length * 2);
        for (byte b : bytes) {
            sb.append(String.format("%02x", b & 0xff));
        }
        return sb.toString();
    }

    /** SDK 版本号 */
    public static String getVersion() {
        return VERSION;
    }
}
